server/vulnerabilities/oval/analyzer.go

```diff
@@ -0,0 +1,175 @@
+package oval
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	oval_parsed "github.com/fleetdm/fleet/v4/server/vulnerabilities/oval/parsed"
+)
+
+// Analyze scans all hosts for vulnerabilities based on the OVAL definitions for their platform.
+func Analyze(
+	ctx context.Context,
+	ds fleet.Datastore,
+	versions *fleet.OSVersions,
+	vulnPath string,
+) ([]fleet.SoftwareVulnerability, error) {
+	for _, v := range versions.OSVersions {
+		platform := NewPlatform(v.Platform, v.Name)
+		if !platform.IsSupported() {
+			continue
+		}
+
+		defs, err := loadDef(platform, vulnPath)
+		if err != nil {
+			return nil, err
+		}
+
+		hIds, err := ds.HostIDsByPlatform(ctx, v.Platform, v.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, hId := range hIds {
+			h := fleet.Host{ID: hId}
+			opts := fleet.SoftwareListOptions{
+				SkipLoadingCVEs: true,
+				VulnerableOnly:  false,
+				WithHostCounts:  false,
+			}
+			err := ds.LoadHostSoftware(ctx, &h, opts)
```

Comment:
Instead of iterating hosts one-by-one (1 Datastore call per Host) shouldn't you just load all software of a given platform with one Datastore call?

E.g. similar to `ListVulnerableSoftwareBySource`, it could be a new method `ListSoftwareByPlatform`.

Reply:
Because when evaluating vulnerabilities we need to look at all software installed on a host as a whole - if I where to load all software at once, there's no way to group the results per host, unless I introduce a new type (a super set with `Software` with a `HostID` prop or something like that, and do the grouping on the server side (AFAIK). 

Plus, until we load test this, I'm not sure what would be more performant, a bunch of small queries or a big query that loads all software.

Reply:
> there's no way to group the results per host

Why do we need to group the vulnerability results here per host? (`ds.DeleteVulnerabilitiesByCPECVE` and `ds.InsertVulnerabilities` do not care about hosts)

--

Am just worried because for every host we are doing two expensive read queries:
1. `ds.LoadHostSoftware`.
2. `ds.ListSoftwareVulnerabilities`.

Reply:
See [this](https://github.com/fleetdm/fleet/pull/5797/files#diff-961c03b44a99f40735d343eff4b67684c8014936b1e007aa794dba2af17360c9R61 ) - when evaluating OVAL definitions we need to look at all installed software as a whole because a criteria can reference two different pieces of software

Reply:
Aha! Now it makes sense. Thanks!

Reply:
Correct me @juan-fdz-hawa if I'm wrong, but the reason we are doing per host is that it's how OVAL works given that rules are not isolated per software but by conditions in the host. So a CVE might be found if a number of conditions are true within the scope of the host.

So if we remove the concept of the host here, we basically cannot apply the OVAL rules.

Reply:
Thanks folks, much more clear now.